### ISSUE TITLE
Incorrect suggestion: use section for substraction and division

### ISSUE THREAD

**Issue by wwmoraes** [ndmitchell/hlint#1466]
Given the functions:

```haskell
op1 :: [Integer] -> Integer
op1 = product . map ((-) 2)

op2 :: [Integer] -> Integer
op2 = product . map ((/) 2)
```

hlint reports:

```text
Suggestion: Use section
Found:
  ((-) 2)
Perhaps:
  (2 -)

Suggestion: Use section
Found:
  ((/) 2)
Perhaps:
  (2 /)
```

Both substraction and division are non-commutative operations, so the suggestions changes both functions' semantics.

**Comment by shayne-fletcher** [comment]
try this:
```
f :: Integer -> Integer
f = ((-) (2 :: Integer))

main = do
  putStrLn $ show(f 0)
  putStrLn $ show(f 1)
  putStrLn $ show(f 2)
  putStrLn $ show(f 3)
  putStrLn $ show(f 4)
```
this should show that indeed `f` is equivalent to the section `(2 -)`. here `(-)` is the binary subtraction operator in prefix form and `(-) 2` a partial application.

**Comment by shayne-fletcher** [comment]
`subtract` can be used in place of the disallowed section (https://hackage.haskell.org/package/base-4.18.0.0/docs/Prelude.html#v:subtract).

**Comment by ndmitchell** [comment]
I think in all these cases the suggestion is indeed equivalent to what you wrote, but also quite a bit clearer - it makes it clear you aren't dividing these values by 2 (which I guess is what you were aiming for?) but in fact dividing 2 by each integer in the list.

**Comment by shayne-fletcher** [comment]
i think ``(`subtract` 2)`` & ``(`div` 2)`` are indeed the sections that were intended.

**Comment by ndmitchell** [comment]
`div` and `/` are different functions. `(- 2)` and `(subtract 2)` are equivalent, but not the same as `(2 -)` which is what was written here.

**Comment by shayne-fletcher** [comment]
> `div` and `/` are different functions. 

yes

> `(- 2)` and `(subtract 2)` are equivalent, but not the same as `(2 -)` which is what was written here.

we are saying the same thing. i am also saying that (note infix ticks) ``(`subtract` 2)`` is `\x -> x - 2`.

**Comment by wwmoraes** [comment]
> i think `` (`subtract` 2) `` & `` (`div` 2) `` are indeed the sections that were intended.

Precisely. However I think my issue now is beyond hlint specifically:

```haskell
ghci> (> 2) (3 :: Integer)
True
ghci> ((-) 2) 3 :: Integer
-1
ghci> (`subtract` 2) 3 :: Integer
-1
ghci> (\x -> x - 2) 3 :: Integer
1
ghci> (flip (-) 2) 3 :: Integer
1
```

Which makes sense on the `(-)` case as it is coerced to a prefix form. Now even an infix partial application of `substract` does not bind the `2` as the second parameter, which is ~the expected behaviour~ my expectation here. I guess I'll have to stick to the good and old `flip` to make sure 😅 

Should hlint instead say that this partial application does not work as expected if the user expects the same semantics of comparison operators? If not, then I'm fine with closing this.

By the way, outstanding work with this tool @ndmitchell, it speeds up my Haskell learning by leaps and bounds! 🎉

**Comment by ndmitchell** [comment]
> Should hlint instead say that this partial application does not work as expected if the user wants the same semantics of comparison operators?

I'm confused. You wrote `((-) 2)`. HLint suggested you could write `(2 -)` instead, which is indeed equivalent. Those things are indeed equivalent. What are you wondering that HLint should recommend or warn instead?

If you actually want `flip (-) 2` then `subtract 2` (no backticks) does it fine. If it were anything other than `-` then `(- 2)` would work, but in Haskell thats a negative number.

**Comment by shayne-fletcher** [comment]
i was completely wrong about ``(`subtract` 2)`` (sorry!). absolutely, `(subtract 2)` works fine.

**Comment by wwmoraes** [comment]
> > Should hlint instead say that this partial application does not work as expected if the user wants the same semantics of comparison operators?
> 
> I'm confused. You wrote `((-) 2)`. HLint suggested you could write `(2 -)` instead, which is indeed equivalent. Those things are indeed equivalent. What are you wondering that HLint should recommend or warn instead?

Sorry to confuse you, I was confused initially myself 🤣 

> If you actually want `flip (-) 2` then `subtract 2` (no backticks) does it fine. If it were anything other than `-` then `(- 2)` would work, but in Haskell thats a negative number.

Thanks, `subtract` is the solution for me. I expected `(-) 2` to behave like `> 2` and bind 2 as the second parameter, because `- 2` alone, as you mentioned, is parsed as a negative number. However that conflicts with the prefix coercion, so quite a corner case that `subtract` solves.

As for hlint, the current result is correct in this case; so my suggestion is, if we consider my confusion something worth warning, then perhaps a mention to use either `2 -` or `subtract 2`, depending on what the user wants, and mention the exception on its binding semantics. Again, either way I'm fine.

Thank you both, I've learned a lot with this discussion!

**Comment by shayne-fletcher** [comment]
haskell report documents all this, https://www.haskell.org/onlinereport/haskell2010/haskellch3.html#x8-280003.4. suggest closing is the right action
